### PR TITLE
[SPARK-7555][docs] Add doc for elastic net in ml-guide and mllib-guide

### PR DESCRIPTION
**Pull request by coderxiang**
@jkbradley I put the elastic net under the **Algorithm guide** section. Also add the formula of elastic net in mllib-linear `mllib-linear-methods#regularizers`. 

@dbtsai I left the code tab for you to add example code. Do you think it is the right place?

### CODE REVIEW

File: docs/ml-guide.md [hunk 1]
```diff
@@ -157,6 +174,49 @@ There are now several algorithms in the Pipelines API which are not in the lower
 * [Feature Extraction, Transformation, and Selection](ml-features.html)
 * [Ensembles](ml-ensembles.html)
 
+## Linear Methods with Elastic Net Regularization
+
+In MLlib, we implement popular linear methods such as logistic regression and linear least squares with L1 or L2 regularization. Refer to [the linear methods section](mllib-linear-methods.html) for details. In `spark.ml`, we add the [Elastic net](http://users.stat.umn.edu/~zouxx019/Papers/elasticnet.pdf), which is a hybrid of L1 and L2 regularization. Mathematically it is defined as a linear combination of the L1-norm and the L2-norm:
```
Comment (jkbradley) [inline docs/ml-guide.md]: > In `spark.ml`, we add the [Elastic net](http://users.stat.umn.edu/~zouxx019/Papers/elasticnet.pdf), which is a hybrid of L1 and L2 regularization.

This makes it sound like it's only in spark.ml.  Can this please instead say that we provide a Pipelines API?  The main thing needed here is a code example, which should demonstrate how to do L1, L2, and a mix.  (But I like the note about how it uses a different optimizer.)

**Comment by jkbradley** [comment]
@coderxiang Thanks for the doc!

In addition to the link to the Elastic Net paper, could you please add a link to the Wikipedia page [http://en.wikipedia.org/wiki/Elastic_net_regularization]?

I'll try generating the doc once the code examples are there.

**Comment by jkbradley** [comment]
I'm OK with using the Algorithm guide section for now; we can create a separate page later on.

**Comment by coderxiang** [comment]
@jkbradley thanks for the comments. I highlighted the Pipelines API in the revision and include the wiki page.

**Comment by dbtsai** [comment]
Agreed. We should create a separate page for this. For example code, in https://github.com/apache/spark/pull/6576/files , I have full running apps in scala, but this may not be suitable here since in the documentation, we probably only have to provide a code snippet to demonstrate the apis. @coderxiang can you add scala example based on my PR? Thanks.

**Comment by coderxiang** [comment]
@dbtsai sounds good. Let me start the your Scala code.

**Comment by jkbradley** [comment]
@coderxiang One more comment: Could you please mention the phrases "Lasso" and "Ridge Regression" in the spark.ml doc?  E.g., say how to set elasticNetParam to get Lasso and Ridge Regression.  That might help users who are not familiar with Elastic Net.

**Comment by coderxiang** [comment]
@jkbradley @dbtsai I removed the pipeline and `o.a.s.example` part in @dbtsai's code so that the code example can be paste and run directly in spark-shell. Please let me know your thoughts on this. I can add them back if needed.

Also update the definition of elastic net using only one lambda.

**Comment by coderxiang** [comment]
jenkins test this please

File: docs/ml-guide.md [hunk 2]
```diff
@@ -157,6 +174,80 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+**Examples**
+
+<div class="codetabs">
+
+<div data-lang="scala" markdown="1">
+The following code illustrates how to load a sample dataset and use logistic regression with elastic net regularization to fit a model.
+
+{% highlight scala %}
+
+import scala.collection.mutable
```
Comment (jkbradley) [inline docs/ml-guide.md]: Do you need these 2 imports??  I think you can remove these 2 + others below.  Can you please see how many can be removed and still be able to run the example in the Spark shell?

File: docs/ml-guide.md [hunk 3]
```diff
@@ -157,6 +174,80 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+import scala.language.reflectiveCalls
+
+import org.apache.spark.{SparkConf, SparkContext}
```
Comment (jkbradley) [inline docs/ml-guide.md]: No need to have this import in this example (not needed to paste into spark shell)

File: docs/ml-guide.md [hunk 4]
```diff
@@ -157,6 +174,80 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+import org.apache.spark.ml.{Pipeline, PipelineStage}
+import org.apache.spark.ml.classification.{LogisticRegression, LogisticRegressionModel}
+import org.apache.spark.ml.feature.StringIndexer
+import org.apache.spark.mllib.util.MLUtils
+import org.apache.spark.sql.DataFrame
+
+val regParam = 0.3
```
Comment (jkbradley) [inline docs/ml-guide.md]: For a simple example, I would not define extra vals here.  Just put the value in the setter method below.

File: docs/ml-guide.md [hunk 5]
```diff
@@ -157,6 +174,80 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+val elasticNetParam = 0.8
+val tol = 1E-6
+val dataPath = "data/mllib/sample_libsvm_data.txt"
+
+println(s"LogisticRegressionExample with regParam $regParam and elasticNetParam $elasticNetParam")
```
Comment (jkbradley) [inline docs/ml-guide.md]: remove

**Comment by jkbradley** [comment]
I'm trying to keep the example as short & simple as possible since it's for the user guide.

Will you be able to add Java and Python examples too?  Thanks!

**Comment by jkbradley** [comment]
Nice, it's much simpler now.  The current PR looks good except for the Java/Python examples.  Who is going to add them?  If you need me to, then I can send an update to this PR.

**Comment by coderxiang** [comment]
@jkbradley I was planning to reuse the code from the code example that @dbtsai is working on. But feel free to update this PR, especially for the Python code.

**Comment by coderxiang** [comment]
@jkbradley had a offline discuss with @dbtsai, I can work on the the Java and python code example. What do you think?

**Comment by dbtsai** [comment]
What example do you want to work on? Example in the documentation or in the example code base?

**Comment by coderxiang** [comment]
@dbtsai the example code in this user guide.

**Comment by coderxiang** [comment]
jenkins test this please

**Comment by coderxiang** [comment]
@dbtsai @jkbradley finished the Java and Python sample code.

File: docs/ml-guide.md [hunk 6]
```diff
@@ -157,6 +174,118 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+`\[
+\alpha \|\wv\|_1 + (1-\alpha) \frac{1}{2}\|\wv\|_2^2, \alpha \in [0, 1].
+\]`
+By setting $\alpha$ properly, it contains both L1 and L2 regularization as special cases. For example, if a [linear regression](/api/scala/index.html#org.apache.spark.ml.regression.LinearRegression) model is trained with the elastic net parameter $\alpha$ set to $1$, it is equivalent to a [Lasso](http://en.wikipedia.org/wiki/Least_squares#Lasso_method) model. On the other hand, if $\alpha$ is set to $0$, the trained model reduces to a [ridge regression](http://en.wikipedia.org/wiki/Tikhonov_regularization) model. We implement Pipelines API for both linear regression and logistic regression with elastic net regularization.
```
Comment (jkbradley) [inline docs/ml-guide.md]: It's odd to link to the Scala API for linear regression but Wikipedia for other models.  How about all of these links go to Wikipedia (following what the spark.mllib guide does)?

File: docs/ml-guide.md [hunk 7]
```diff
@@ -157,6 +174,118 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+
+// Print the weights and intercept for logistic regression
+println(s"Weights: ${lrModel.weights} Intercept: ${lrModel.intercept}")
+
+{% endhighlight %}
+
+</div>
+
+<div data-lang="java" markdown="1">
+The following code illustrates how to load a sample dataset and use logistic regression with elastic net regularization to fit a model.
```
Comment (jkbradley) [inline docs/ml-guide.md]: This text is not language-specific.  Can it be moved before the code tabs?

File: docs/ml-guide.md [hunk 8]
```diff
@@ -157,6 +174,118 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+
+    // Load training data
+    DataFrame training = sql.createDataFrame(MLUtils.loadLibSVMFile(sc, path).toJavaRDD(), LabeledPoint.class);
+
+    LogisticRegression lr = new LogisticRegression()
+      .setMaxIter(10)
+      .setRegParam(0.3)
+      .setElasticNetParam(0.8)
+      .setThreshold(0.6)
+      .setProbabilityCol("myProbability");
```
Comment (jkbradley) [inline docs/ml-guide.md]: I'd prefer for the examples in different languages to be as similar as possible.  Can you remove the parameters here which are not set in the other languages' examples?

File: docs/ml-guide.md [hunk 9]
```diff
@@ -157,6 +174,118 @@ There are now several algorithms in the Pipelines API which are not in the lower
...
+{% highlight python %}
+
+from pyspark.ml.classification import LogisticRegression
+from pyspark.mllib.regression import LabeledPoint
+from pyspark.mllib.util import MLUtils
+
+# Load training data
+training = MLUtils.loadLibSVMFile(sc, "data/mllib/sample_libsvm_data.txt").toDF()
+
+lr = LogisticRegression(maxIter=10, regParam=0.3)
```
Comment (jkbradley) [inline docs/ml-guide.md]: Let's be Pythonic and set the elasticNetParam via an argument, like the other params.

**Comment by jkbradley** [comment]
@coderxiang My apologies for the long delay!  It looks good, save for the minor comments above.  Since there are merge issues, would you mind going ahead and moving this doc to a new Markdown file, linked from the algorithm section?  That new file can be for Linear Methods and be called ml-linear-methods.md

Thank you!

**Comment by coderxiang** [comment]
@jkbradley thanks for the comments! Just uploaded the change.

**Comment by coderxiang** [comment]
@jkbradley is this one ready to go?

**Comment by jkbradley** [comment]
LGTM merging with master
Thanks for the PR!

**Pull request closed without merging**